### PR TITLE
driver/motor: add upperhalf structure

### PR DESCRIPTION
**Pull request by Donny9**
## Summary
driver/motor: add upperhalf structure and optimize code.
## Impact
optimize code
## Testing
daily test

### CODE REVIEW

File: drivers/motor/foc/foc_dev.c [hunk 1]
```diff
@@ -147,19 +142,13 @@ static int foc_open(FAR struct file *filep)
 
           if (tmp == 1)
             {
-              /* Yes.. perform one time driver setup */
-
-              flags = enter_critical_section();
-
```
Comment (acassis) [inline drivers/motor/foc/foc_dev.c:153]: @xiaoxiang781216 @Donny9 why the critical section is not necessary anymore? Should this be done by the lower half of the driver?

Comment (xiaoxiang781216) [reply drivers/motor/foc/foc_dev.c:153]: The code is already locked by nxsem_wait.

Comment (Donny9) [reply drivers/motor/foc/foc_dev.c:153]: yes, I also think  the critical section should be done by lower half of the driver, more detailed control.

File: drivers/motor/foc/foc_dev.c [hunk 2]
```diff
@@ -862,19 +830,14 @@ int foc_register(FAR const char *path, FAR struct foc_dev_s *dev)
...
       goto errout;
     }
 
   /* Reset counter */
 
   dev->ocount = 0;
 
-  /* Store device number */
-
-  dev->devno = g_devno_cntr;
```
[the remark this answers is not in the document]
Comment (raiden00pl) [reply drivers/motor/foc/foc_dev.c:876]: And it makes the condition above is useless

File: drivers/motor/foc/foc_dev.c (hunk 1 above)
Comment (xiaoxiang781216) [reply drivers/motor/foc/foc_dev.c:153]: Sure, let's fix it one by one.

File: drivers/motor/foc/foc_dev.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply drivers/motor/foc/foc_dev.c:876]: Done.

**Comment by acassis** [comment]
Please rebase to fix the issue in the simulator

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by raiden00pl — COMMENTED** [review]
lgtm

**Pull request merged**